### PR TITLE
return hub exception only from ToIotHubClientContract methods

### PR DESCRIPTION
**Pull request by brycewang-microsoft**
In the AMQP layer, this PR is for both the hub device and service.

### CODE REVIEW

File: iothub/device/src/Transport/AmqpIot/AmqpIotErrorAdapter.cs [hunk 1]
```diff
@@ -52,142 +52,113 @@ public static Exception GetExceptionFromOutcome(Outcome outcome)
...
-            {
-                return new InvalidOperationException(message, amqpException);
-            }
-            else if (Equals(AmqpErrorCode.InvalidField, amqpSymbol))
-            {
-                return new InvalidOperationException(message, amqpException);
-            }
-            else if (Equals(AmqpErrorCode.NotImplemented, amqpSymbol))
-            {
-                return new NotSupportedException(message, amqpException);
```
Comment (brycewang-microsoft) [inline iothub/device/src/Transport/AmqpIot/AmqpIotErrorAdapter.cs:92]: these will be addressed in the "else" block below and returned as `IotHubClientException`.

File: iothub/service/src/Amqp/AmqpClientHelper.cs [hunk 2]
```diff
@@ -114,69 +109,52 @@ internal static Exception ToIotHubClientContract(Error error)
...
             {
-                retException = new IotHubServiceException(message, HttpStatusCode.NotFound, IotHubServiceErrorCode.DeviceNotFound);
-            }
-            else if (error.Condition.Equals(AmqpErrorCode.NotImplemented))
-            {
-                retException = new NotSupportedException(message);
-            }
-            else if (error.Condition.Equals(AmqpErrorCode.NotAllowed))
-            {
-                retException = new InvalidOperationException(message);
```
Comment (brycewang-microsoft) [inline iothub/service/src/Amqp/AmqpClientHelper.cs:129]: similarly here, these will be addressed in the "else" block below and returned as `IotHubServiceException`.

File: iothub/service/src/Amqp/AmqpClientHelper.cs [hunk 3]
```diff
@@ -114,69 +109,52 @@ internal static Exception ToIotHubClientContract(Error error)
...
             {
-                retException = new IotHubServiceException(message, HttpStatusCode.Unauthorized, IotHubServiceErrorCode.IotHubUnauthorizedAccess);
-            }
-            else if (error.Condition.Equals(IotHubAmqpErrorCode.ArgumentError))
-            {
-                retException = new ArgumentException(message);
-            }
-            else if (error.Condition.Equals(IotHubAmqpErrorCode.ArgumentOutOfRangeError))
-            {
-                retException = new ArgumentOutOfRangeException(message);
```
Comment (brycewang-microsoft) [inline iothub/service/src/Amqp/AmqpClientHelper.cs:141]: same thing here

File: iothub/device/src/Transport/AmqpIot/AmqpIotErrorAdapter.cs [hunk 4]
```diff
@@ -52,142 +52,113 @@ public static Exception GetExceptionFromOutcome(Outcome outcome)
...
             {
-                return new IotHubClientException(message, true, amqpException);
+                retException = new(message, true, amqpException);
             }
             else if (Equals(AmqpErrorCode.Stolen, amqpSymbol))
             {
-                return new IotHubClientException(message, true, amqpException);
+                retException = new(message, true, amqpException);
             }
             // AMQP Transaction Error
```
Comment (drwill-ms) [inline iothub/device/src/Transport/AmqpIot/AmqpIotErrorAdapter.cs:139]: Does this comment need to move or be removed?

File: iothub/device/src/Transport/AmqpIot/AmqpIotErrorAdapter.cs [hunk 5]
```diff
@@ -252,21 +207,20 @@ public static Exception ToIotHubClientContract(Error error)
             }
             else if (error.Condition.Equals(DeviceContainerThrottled))
             {
-                retException = new IotHubClientException(message, IotHubClientErrorCode.Throttled);
+                retException = new(message, IotHubClientErrorCode.Throttled);
             }
             else if (error.Condition.Equals(IotHubSuspended))
             {
-                retException = new IotHubClientException("Iothub {0} is suspended".FormatInvariant(message), IotHubClientErrorCode.Suspended);
+                retException = new("Iothub {0} is suspended".FormatInvariant(message), IotHubClientErrorCode.Suspended);
```
Comment (drwill-ms) [inline iothub/device/src/Transport/AmqpIot/AmqpIotErrorAdapter.cs:214]:
```suggestion
                retException = new("IoT hub {0} is suspended".FormatInvariant(message), IotHubClientErrorCode.Suspended);
```

File: iothub/service/src/Amqp/AmqpClientHelper.cs [hunk 6]
```diff
@@ -36,18 +36,13 @@ internal static Exception ToIotHubClientContract(Exception exception)
...
-                        // pass amqpException as the inner exception of IotHubServiceException
-                        return new IotHubServiceException(
-                            hubEx.Message,
-                            hubEx.StatusCode,
-                            hubEx.ErrorCode,
-                            string.Empty,
-                            amqpException);
-                    }
-                    return ex;
+                    IotHubServiceException ex = ToIotHubClientContract(amqpException.Error);
```
Comment (drwill-ms) [inline iothub/service/src/Amqp/AmqpClientHelper.cs:39]: Bummer that we have to create a whole new exception when we have one. Perhaps we pass in the inner to ToIotHubClientContract?

**Review by drwill-ms — APPROVED** [review]

**Comment by drwill-ms** [comment]
Please get sign-off from Tim before merge.

File: iothub/device/src/Transport/AmqpIot/AmqpIotErrorAdapter.cs (hunk 4 above)
Comment (brycewang-microsoft) [reply iothub/device/src/Transport/AmqpIot/AmqpIotErrorAdapter.cs:139]: It seems that this comment is for several transaction-error-related AmqpErrorCode in the following else-ifs, so no, we can just keep it as-is.

File: iothub/service/src/Amqp/AmqpClientHelper.cs (hunk 6 above)
Comment (brycewang-microsoft) [reply iothub/service/src/Amqp/AmqpClientHelper.cs:39]: Sure, I like this idea!

File: iothub/device/src/Transport/AmqpIot/AmqpIotErrorAdapter.cs [hunk 7]
```diff
@@ -52,142 +52,113 @@ public static Exception GetExceptionFromOutcome(Outcome outcome)
...
             AmqpSymbol amqpSymbol = error.Condition;
             string message = error.ToString();
 
+            IotHubClientException retException;
+
             // Generic AMQP error
             if (Equals(AmqpErrorCode.InternalError, amqpSymbol))
             {
-                return new IotHubClientException(message, IotHubClientErrorCode.NetworkErrors, amqpException);
+                retException = new(message, IotHubClientErrorCode.NetworkErrors, amqpException);
```
Comment (timtay-microsoft) [inline iothub/device/src/Transport/AmqpIot/AmqpIotErrorAdapter.cs:66]: @drwill-ms how do you feel about using ```new()``` in contexts like these where the variable's type isn't immediately obvious?

Should it be
```csharp
retException = new IotHubClientException(...);
```
instead?

Comment (drwill-ms) [reply iothub/device/src/Transport/AmqpIot/AmqpIotErrorAdapter.cs:66]: Yeah, we should basically just use it for class fields where the type must be on the left or for inline init of properties. I agree, in this case it isn't as obvious so we should be explicit.

**Review by timtay-microsoft — APPROVED** [review]
Just the one comment. Otherwise looks good

**Pull request merged**